src/pudl_archiver/zenodo/api_client.py

```diff
@@ -0,0 +1,469 @@
+                    exclude={"publication_date", "doi", "prereserve_doi"},
+                )
+            }
+        )
+
+        async with self.session.post(
+            url, params=params, data=data, headers=headers
+        ) as response:
+            # Ignore content type
+            return Deposition(**await response.json(content_type=None))
+
+    async def add_files(self, resources: dict[str, ResourceInfo]):
+        """Add all downloaded files to zenodo deposition.
+
+        This method will check local files against remote files which already
+        exist in Zenodo deposition. If local file name matches remote file name,
+        checksums will be used to determine if file has changed. It will also
+        create and upload a new datapackage.
+
+        Args:
+            resources: Dictionary mapping filenames to ResourceInfo objects,
+            which contain local path to resource and working partitions.
+
+        Returns:
+            Updated Deposition object.
+        """
+        for name, resource in resources.items():
+            if name not in self.deposition_files:
+                self.changed = True
+                logger.info(f"Adding {name} to deposition.")
+
+                with open(resource.local_path, "rb") as f:
+                    await self.upload(f, resource.local_path.name)
+            else:
+                file_info = self.deposition_files[name]
+
+                # If file is not exact match for existing file, update with new file
+                if not _compute_md5(resource.local_path) == file_info.checksum:
+                    self.changed = True
+                    logger.info(f"Updating {name}")
+                    await self.delete_file(file_info)
+
+                    with open(resource.local_path, "rb") as f:
+                        await self.upload(f, resource.local_path.name)
+
+        # Delete files not included in new deposition
+        for filename, file in self.deposition_files.items():
+            if filename not in resources and filename != "datapackage.json":
+                self.changed = True
+                await self.delete_file(file)
+
+        await self.update_datapackage(resources)
+
+    async def get_deposition(self, concept_doi: str):
+        """Get data for a single Zenodo Deposition based on the provided query.
+
+        See https://developers.zenodo.org for more information.
+
+        Args:
+            concept_doi: Concept DOI for desired data source.
+
+        Returns:
+            Deposition metadata pertaining to the latest version returned by Zenodo api.
+        """
+        url = f"{self.api_root}/deposit/depositions"
+        params = {"q": f'conceptdoi:"{concept_doi}"', "access_token": self.upload_key}
+
+        async with self.session.get(url, params=params) as response:
+            # Deposition should be first element in json response
+            raw_json = await response.json()
+
+            return Deposition(**raw_json[0])
```

Comment:
Can we raise an exception here if the first element of `raw_json` doesn't match our expectations?

Reply:
Added a check for the array length, and I think any other errors will get caught in the pydantic validation of the `Deposition` object

Reply:
Oh that's right. Thanks pydantic.